File: frontend/src/api/live.ts
import { z } from 'zod';
import { ApiError, API_BASE, getJson, postJson } from './client';
import { analysisResponseSchema } from './analysis';

/**
 * Live Intelligence transport (Phase 13 Part 2A). Simulated history only.
 *
 * ## Every response is checked for what it claims to be
 *
 * Provenance and market currency are `z.literal`s. A response that called
 * itself an exchange feed, or a current market, would fail to parse and never
 * reach application state - the browser does not relabel data, and it does
 * not accept data that relabels itself.
 *
 * ## The browser sends intent, never facts
 *
 * The only request bodies here name a stored dataset, timeframes, a window and
 * a playback pace - or ask for an analysis of what the server already holds.
 * There is no field for a candle, a price, a time, a status or a provenance.
 *
 * ## One place owns the event stream
 *
 * `openLiveEvents` is the only `EventSource` in the application. It closes the
 * stream on any transport error instead of letting the browser reconnect on
 * its own: reconnection is a *resynchronisation* - re-read the authoritative
 * snapshot, then resume from its cursor - and that decision belongs to the
 * screen, not to a retry loop that would silently skip what it missed.
 */

const provenance = z.literal('SIMULATED_HISTORICAL_STREAM');
const currency = z.literal('HISTORICAL');

const limitsSchema = z.object({
  max_sessions: z.number().int(),
  max_subscribers_per_session: z.number().int(),
  max_subscribers_total: z.number().int(),
  subscriber_queue: z.number().int(),
  timeline_retention: z.number().int(),
  timeline_page_max: z.number().int(),
  max_session_seconds: z.number(),
  creations_per_minute: z.number().int(),
  max_concurrent_analyses: z.number().int(),
  min_window_candles: z.number().int(),
  max_window_candles: z.number().int(),
  max_closed_candles: z.number().int(),
  max_missing_sequences: z.number().int(),
  max_recorded_issues: z.number().int(),
  max_reconnects: z.number().int(),
  heartbeat_seconds: z.number(),
});

export const liveCapabilitySchema = z.object({
  state: z.enum(['AVAILABLE', 'DISABLED']),
  detail: z.string(),
  provenance,
  market_currency: currency,
  real_exchange_connected: z.literal(false),
  execution: z.literal('DISABLED'),
  transport: z.literal('SSE'),
  scope: z.literal('LOCAL_DEVELOPMENT'),
  paces: z.array(z.string()),
  limits: limitsSchema.nullable(),
});

export const liveSourceSchema = z.object({
  source_id: z.string(),
  instrument_label: z.string(),
  origin: z.literal('USER_SUPPLIED_HISTORICAL'),
  contract_identity: z.literal('NOT_ESTABLISHED'),
  streamable: z.boolean(),
  refusal: z.string().nullable(),
  timeframes: z.array(
    z.object({
      timeframe: z.string(),
      rows: z.number().int(),
      first_open_time: z.string(),
      last_open_time: z.string(),
    }),
  ),
});

const sourceListSchema = z.object({
  items: z.array(liveSourceSchema),
  total: z.number().int(),
  offset: z.number().int(),
  limit: z.number().int(),
});

const candleSchema = z.object({
  state: z.enum(['CLOSED', 'FORMING']),
  market_open_time: z.string(),
  market_coverage_end: z.string(),
  market_event_time: z.string(),
  received_at: z.string(),
  open: z.string(),
  high: z.string(),
  low: z.string(),
  close: z.string(),
  volume: z.string(),
  sequence: z.number().int().nullable(),
});

export const FRESHNESS = ['NO_DATA', 'FRESH', 'STALE'] as const;
export const INTEGRITY = [
  'COMPLETE',
  'GAPPED',
  'DISCONTINUOUS',
  'CONFLICTED',
  'UNVERIFIED',
] as const;
export const CONNECTION = [
  'INITIALIZING',
  'CONNECTED',
  'DISCONNECTED',
  'RECOVERING',
  'TERMINATED',
] as const;

const timeframeSchema = z.object({
  timeframe: z.string(),
  freshness: z.enum(FRESHNESS),
  freshness_threshold_seconds: z.number(),
  integrity: z.enum(INTEGRITY),
  availability: z.enum(['AVAILABLE', 'UNAVAILABLE']),
  reasons: z.array(z.string()),
  closed_count: z.number().int(),
  trimmed: z.number().int(),
  first_closed_open_time: z.string().nullable(),
  last_closed_open_time: z.string().nullable(),
  last_sequence: z.number().int().nullable(),
  last_market_event_time: z.string().nullable(),
  last_received_at: z.string().nullable(),
  missing_sequences: z.number().int(),
  missing_overflowed: z.boolean(),
  temporal_gaps: z.number().int(),
  sequence_mismatches: z.number().int(),
  conflicts: z.number().int(),
  duplicates: z.number().int(),
  late_fills: z.number().int(),
  awaiting_continuity: z.boolean(),
  forming_ahead: z.boolean(),
  unresolved_trimmed: z.number().int(),
  version: z.number().int(),
  latest_confirmed: candleSchema.nullable(),
  forming: candleSchema.nullable(),
});

export const liveSessionSchema = z.object({
  id: z.string(),
  simulated: z.literal(true),
  execution: z.literal('DISABLED'),
  identity: z.object({
    stream_id: z.string(),
    source_id: z.string(),
    source_origin: z.literal('USER_SUPPLIED_HISTORICAL'),
    provider_id: z.string(),
    instrument_label: z.string(),
    contract_identity: z.literal('NOT_ESTABLISHED'),
    provenance,
    market_currency: currency,
  }),
  lifecycle: z.enum(['RUNNING', 'ENDED']),
  end_origin: z.enum(['STREAM', 'USER_CANCELLED', 'DEADLINE', 'SHUTDOWN']).nullable(),
  connection: z.enum(CONNECTION),
  termination_reason: z.string().nullable(),
  reconnects: z.number().int(),
  created_at: z.string(),
  ended_at: z.string().nullable(),
  snapshot_at: z.string(),
  playback: z.object({
    pace: z.string(),
    event_spacing_seconds: z.number(),
    total_events: z.number().int(),
    market_window_start: z.string(),
    market_window_end: z.string(),
    candles_per_timeframe: z.record(z.number().int()),
    forming_candles_published: z.literal(false),
  }),
  timeframes: z.array(timeframeSchema),
  alerts: z.array(
    z.object({ kind: z.string(), timeframe: z.string().nullable(), detail: z.string() }),
  ),
  rejection_counts: z.record(z.number().int()),
  recent_rejections: z.array(
    z.object({
      timeframe: z.string().nullable(),
      code: z.string(),
      detail: z.string(),
      received_at: z.string(),
    }),
  ),
  analysis: z.object({
    analyses_run: z.number().int(),
    available_timeframes: z.array(z.string()),
    last_market_as_of: z.string().nullable(),
    last_requested_at: z.string().nullable(),
    last_current: z.boolean(),
  }),
  cursor: z.number().int(),
  oldest_retained: z.number().int(),
  subscribers: z.number().int(),
});

const sessionListSchema = z.object({
  items: z.array(
    z.object({
      id: z.string(),
      instrument_label: z.string(),
      source_id: z.string(),
      lifecycle: z.string(),
      connection: z.string(),
      created_at: z.string(),
      cursor: z.number().int(),
    }),
  ),
  capacity: z.number().int(),
});

export const timelineEntrySchema = z.object({
  seq: z.number().int(),
  kind: z.string(),
  recorded_at: z.string(),
  code: z.string(),
  timeframe: z.string().nullable(),
  market_open_time: z.string().nullable(),
  market_event_time: z.string().nullable(),
  sequence: z.number().int().nullable(),
  before: z.string().nullable(),
  after: z.string().nullable(),
  backfill: z.boolean(),
});

const timelinePageSchema = z.object({
  session_id: z.string(),
  entries: z.array(timelineEntrySchema),
  cursor: z.number().int(),
  oldest_retained: z.number().int(),
  gap: z.boolean(),
  limit: z.number().int(),
});

export const liveAnalysisSchema = z.object({
  session_id: z.string(),
  provenance,
  market_currency: currency,
  market_as_of: z.string(),
  requested_at: z.string(),
  included: z.array(z.string()),
  excluded: z.array(z.object({ timeframe: z.string(), reasons: z.array(z.string()) })),
  fingerprint: z.array(z.object({ timeframe: z.string(), version: z.number().int() })),
  reused: z.boolean(),
  current: z.boolean(),
  analysis: analysisResponseSchema,
  session: liveSessionSchema,
});

export const liveEnvelopeSchema = z.object({
  protocol: z.literal('viop.live.v1'),
  session_id: z.string(),
  event_id: z.number().int().nullable(),
  kind: z.enum(['TIMELINE', 'STATE', 'RESYNC_REQUIRED', 'END', 'HEARTBEAT']),
  server_time: z.string(),
  cursor: z.number().int(),
  entry: timelineEntrySchema.nullable(),
  session: liveSessionSchema.nullable(),
  reason: z.string().nullable(),
});

export type LiveCapabilityDto = z.infer<typeof liveCapabilitySchema>;
export type LiveSourceDto = z.infer<typeof liveSourceSchema>;
export type LiveSessionDto = z.infer<typeof liveSessionSchema>;
export type LiveTimeframeDto = LiveSessionDto['timeframes'][number];
export type LiveCandleDto = z.infer<typeof candleSchema>;
export type LiveSessionListDto = z.infer<typeof sessionListSchema>;
export type LiveTimelineEntryDto = z.infer<typeof timelineEntrySchema>;
export type LiveTimelinePageDto = z.infer<typeof timelinePageSchema>;
export type LiveAnalysisDto = z.infer<typeof liveAnalysisSchema>;
export type LiveEnvelopeDto = z.infer<typeof liveEnvelopeSchema>;

/** What a person chooses. Nothing here is a market fact. */
export interface CreateLivePayload {
  readonly source_id: string;
  readonly timeframes: readonly string[];
  readonly window_candles: number;
  readonly pace: 'SLOW' | 'NORMAL' | 'FAST';
}

const LIVE = '/live';

function sessionPath(sessionId: string): string {
  return `${LIVE}/sessions/${encodeURIComponent(sessionId)}`;
}

export function getLiveCapability(signal?: AbortSignal): Promise<LiveCapabilityDto> {
  return getJson(`${LIVE}/capability`, liveCapabilitySchema, signal ? { signal } : undefined);
}

export function listLiveSources(signal?: AbortSignal) {
  return getJson(`${LIVE}/sources`, sourceListSchema, signal ? { signal } : undefined);
}

export function listLiveSessions(signal?: AbortSignal): Promise<LiveSessionListDto> {
  return getJson(`${LIVE}/sessions`, sessionListSchema, signal ? { signal } : undefined);
}

export function createLiveSession(payload: CreateLivePayload): Promise<LiveSessionDto> {
  return postJson(`${LIVE}/sessions`, payload, liveSessionSchema);
}

export function getLiveSession(sessionId: string, signal?: AbortSignal): Promise<LiveSessionDto> {
  return getJson(sessionPath(sessionId), liveSessionSchema, signal ? { signal } : undefined);
}

export function cancelLiveSession(sessionId: string): Promise<LiveSessionDto> {
  return postJson(`${sessionPath(sessionId)}/cancel`, {}, liveSessionSchema);
}

export async function removeLiveSession(sessionId: string): Promise<void> {
  let response: Response;
  try {
    response = await fetch(`${API_BASE}${sessionPath(sessionId)}`, { method: 'DELETE' });
  } catch (cause) {
    throw new ApiError(cause instanceof Error ? cause.message : 'network error', null);
  }
  // 404 means it is already gone - the outcome the caller asked for.
  if (!response.ok && response.status !== 404) {
    throw new ApiError(`request failed (${response.status})`, response.status);
  }
}

export function readLiveTimeline(
  sessionId: string,
  after: number | null,
  signal?: AbortSignal,
): Promise<LiveTimelinePageDto> {
  const query = after === null ? '?limit=100' : `?after=${after}&limit=100`;
  return getJson(
    `${sessionPath(sessionId)}/timeline${query}`,
    timelinePageSchema,
    signal ? { signal } : undefined,
  );
}

export function analyseLiveSession(
  sessionId: string,
  signal?: AbortSignal,
): Promise<LiveAnalysisDto> {
  return postJson(
    `${sessionPath(sessionId)}/analysis`,
    {},
    liveAnalysisSchema,
    signal ? { signal } : undefined,
  );
}

export interface LiveEventHandlers {
  readonly onOpen: () => void;
  readonly onEnvelope: (envelope: LiveEnvelopeDto) => void;
  /** The browser's connection to *this server* failed. Not the provider. */
  readonly onTransportLost: () => void;
  readonly onInvalid: () => void;
}

/** Open the session's event stream. Returns the function that closes it. */
export function openLiveEvents(
  sessionId: string,
  after: number | null,
  handlers: LiveEventHandlers,
): () => void {
  const query = after === null ? '' : `?after=${after}`;
  const source = new EventSource(`${API_BASE}${sessionPath(sessionId)}/events${query}`);
  let closed = false;
  const close = () => {
    closed = true;
    source.close();
  };
  source.onopen = () => {
    if (!closed) handlers.onOpen();
  };
  source.onmessage = (event: MessageEvent<string>) => {
    if (closed) return;
    let payload: unknown;
    try {
      payload = JSON.parse(event.data);
    } catch {
      handlers.onInvalid();
      return;
    }
    const parsed = liveEnvelopeSchema.safeParse(payload);
    if (!parsed.success) {
      handlers.onInvalid();
      return;
    }
    handlers.onEnvelope(parsed.data);
  };
  source.onerror = () => {
    if (closed) return;
    // Never let the browser retry on its own: a reconnect must resync.
    close();
    handlers.onTransportLost();
  };
  return close;
}

File: backend/app/api/schemas/live.py
"""Live Intelligence request and response schemas (Phase 13 Part 2A).

## A client supplies intent, never facts

Creating a session names a stored dataset, which timeframes to stream, how
large a window and how fast to play it. There is no field for a candle, a
price, a timestamp, a sequence number, a provenance, a market currency, a
stream status, contract metadata, an analysis, a risk approval, a position or
an order. A body that carries one is a 422 because the field does not exist -
``extra="forbid"`` on every request model.

## Every response says what it is

Provenance and market currency are ``Literal`` fields with exactly one value
each. A response model *cannot* be built claiming exchange data or a current
market, whatever the code upstream does - the validation error would be a 500
before it could be a false label.

## Three clocks, three names

``market_*`` fields are market time (from the stored candles). ``received_at``
and ``*_seconds`` freshness fields are the receive clock. ``created_at``,
``ended_at``, ``snapshot_at``, ``recorded_at`` and ``server_time`` are the
server's audit clock. None is derived from another, and none from a browser.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, StrictInt

from app.api.schemas.analysis import AccountBody, AnalysisResponse, RiskSettingsBody

_PUBLIC = ConfigDict(extra="forbid", frozen=True)

LiveTimeframe = Literal["5M", "15M", "1H", "1D"]
Provenance = Literal["SIMULATED_HISTORICAL_STREAM"]
Currency = Literal["HISTORICAL"]

DATASET_ID_PATTERN = r"^RD-[0-9a-f]{32}$"
SESSION_ID_PATTERN = r"^LS-[0-9a-f]{24}$"


# ----------------------------------------------------------------------
# Requests
# ----------------------------------------------------------------------


class CreateLiveSessionBody(BaseModel):
    """Which stored history to play, and how. Nothing about the market."""

    model_config = _PUBLIC

    source_id: str = Field(pattern=DATASET_ID_PATTERN)
    timeframes: list[LiveTimeframe] = Field(min_length=1, max_length=4)
    window_candles: StrictInt = Field(default=300, ge=50, le=1000)
    """Candles of the finest chosen timeframe; coarser ones cover the same
    market stretch."""

    pace: Literal["SLOW", "NORMAL", "FAST"] = "NORMAL"


class LiveAnalysisBody(BaseModel):
    """Optional sizing inputs. The candles are the session's, not the client's."""

    model_config = _PUBLIC

    account: AccountBody | None = None
    risk: RiskSettingsBody | None = None


# ----------------------------------------------------------------------
# Capability and sources
# ----------------------------------------------------------------------


class LiveLimitsResponse(BaseModel):
    model_config = _PUBLIC

    max_sessions: int
    max_subscribers_per_session: int
    max_subscribers_total: int
    subscriber_queue: int
    timeline_retention: int
    timeline_page_max: int
    max_session_seconds: float
    creations_per_minute: int
    max_concurrent_analyses: int
    min_window_candles: int
    max_window_candles: int
    max_closed_candles: int
    max_missing_sequences: int
    max_recorded_issues: int
    max_reconnects: int
    heartbeat_seconds: float


class LiveCapabilityResponse(BaseModel):
    model_config = _PUBLIC

    state: Literal["AVAILABLE", "DISABLED"]
    detail: str
    provenance: Provenance = "SIMULATED_HISTORICAL_STREAM"
    market_currency: Currency = "HISTORICAL"
    real_exchange_connected: Literal[False] = False
    execution: Literal["DISABLED"] = "DISABLED"
    transport: Literal["SSE"] = "SSE"
    scope: Literal["LOCAL_DEVELOPMENT"] = "LOCAL_DEVELOPMENT"
    paces: list[str]
    limits: LiveLimitsResponse | None


class SourceTimeframeResponse(BaseModel):
    model_config = _PUBLIC

    timeframe: str
    rows: int
    first_open_time: str
    last_open_time: str


class LiveSourceResponse(BaseModel):
    model_config = _PUBLIC

    source_id: str
    instrument_label: str
    """The dataset's symbol text - a label, never a verified contract."""

    origin: Literal["USER_SUPPLIED_HISTORICAL"]
    contract_identity: Literal["NOT_ESTABLISHED"] = "NOT_ESTABLISHED"
    streamable: bool
    refusal: str | None
    timeframes: list[SourceTimeframeResponse]


class LiveSourceListResponse(BaseModel):
    model_config = _PUBLIC

    items: list[LiveSourceResponse]
    total: int
    offset: int
    limit: int


# ----------------------------------------------------------------------
# Sessions
# ----------------------------------------------------------------------


class LiveCandleResponse(BaseModel):
    """One candle as the stream delivered it. Money-shaped values are text."""

    model_config = _PUBLIC

    state: Literal["CLOSED", "FORMING"]
    market_open_time: str
    market_coverage_end: str
    market_event_time: str
    received_at: str
    open: str
    high: str
    low: str
    close: str
    volume: str
    sequence: int | None


class LiveTimeframeResponse(BaseModel):
    model_config = _PUBLIC

    timeframe: str
    freshness: Literal["NO_DATA", "FRESH", "STALE"]
    """Transport freshness on the receive clock. Never market currency."""

    freshness_threshold_seconds: float
    integrity: Literal["COMPLETE", "GAPPED", "DISCONTINUOUS", "CONFLICTED", "UNVERIFIED"]
    availability: Literal["AVAILABLE", "UNAVAILABLE"]
    reasons: list[str]
    closed_count: int
    trimmed: int
    first_closed_open_time: str | None
    last_closed_open_time: str | None
    last_sequence: int | None
    last_market_event_time: str | None
    last_received_at: str | None
    missing_sequences: int
    missing_overflowed: bool
    temporal_gaps: int
    sequence_mismatches: int
    conflicts: int
    duplicates: int
    late_fills: int
    awaiting_continuity: bool
    forming_ahead: bool
    unresolved_trimmed: int
    version: int
    latest_confirmed: LiveCandleResponse | None
    forming: LiveCandleResponse | None


class LiveAlertResponse(BaseModel):
    """A data-quality candidate from Part 1's ``alert_candidates``. Never a
    trade instruction."""

    model_config = _PUBLIC

    kind: str
    timeframe: str | None
    detail: str


class LiveRejectionResponse(BaseModel):
    model_config = _PUBLIC

    timeframe: str | None
    code: str
    detail: str
    received_at: str


class LiveIdentityResponse(BaseModel):
    """Five identities, never merged."""

    model_config = _PUBLIC

    stream_id: str
    source_id: str
    source_origin: Literal["USER_SUPPLIED_HISTORICAL"]
    provider_id: str
    instrument_label: str
    contract_identity: Literal["NOT_ESTABLISHED"] = "NOT_ESTABLISHED"
    provenance: Provenance
    market_currency: Currency


class LivePlaybackResponse(BaseModel):
    model_config = _PUBLIC

    pace: str
    event_spacing_seconds: float
    total_events: int
    market_window_start: str
    market_window_end: str
    candles_per_timeframe: dict[str, int]
    forming_candles_published: Literal[False] = False
    """Stored candles are finished ones; no intermediate state is invented."""


class LiveAnalysisStatusResponse(BaseModel):
    model_config = _PUBLIC

    analyses_run: int
    available_timeframes: list[str]
    last_market_as_of: str | None
    last_requested_at: str | None
    last_current: bool
    """Whether the last analysis still describes this stream's confirmed
    state. Never a claim that the market is current."""


class LiveSessionResponse(BaseModel):
    model_config = _PUBLIC

    id: str
    simulated: Literal[True] = True
    execution: Literal["DISABLED"] = "DISABLED"
    identity: LiveIdentityResponse
    lifecycle: Literal["RUNNING", "ENDED"]
    end_origin: Literal["STREAM", "USER_CANCELLED", "DEADLINE", "SHUTDOWN"] | None
    connection: Literal["INITIALIZING", "CONNECTED", "DISCONNECTED", "RECOVERING", "TERMINATED"]
    termination_reason: str | None
    reconnects: int
    created_at: str
    ended_at: str | None
    snapshot_at: str
    playback: LivePlaybackResponse
    timeframes: list[LiveTimeframeResponse]
    alerts: list[LiveAlertResponse]
    rejection_counts: dict[str, int]
    recent_rejections: list[LiveRejectionResponse]
    analysis: LiveAnalysisStatusResponse
    cursor: int
    oldest_retained: int
    subscribers: int


class LiveSessionSummaryResponse(BaseModel):
    model_config = _PUBLIC

    id: str
    instrument_label: str
    source_id: str
    lifecycle: str
    connection: str
    created_at: str
    cursor: int


class LiveSessionListResponse(BaseModel):
    model_config = _PUBLIC

    items: list[LiveSessionSummaryResponse]
    capacity: int


# ----------------------------------------------------------------------
# Timeline
# ----------------------------------------------------------------------


class LiveTimelineEntryResponse(BaseModel):
    model_config = _PUBLIC

    seq: int
    kind: str
    recorded_at: str
    code: str
    timeframe: str | None
    market_open_time: str | None
    market_event_time: str | None
    sequence: int | None
    before: str | None
    after: str | None
    backfill: bool


class LiveTimelinePageResponse(BaseModel):
    model_config = _PUBLIC

    session_id: str
    entries: list[LiveTimelineEntryResponse]
    cursor: int
    oldest_retained: int
    gap: bool
    limit: int


# ----------------------------------------------------------------------
# Analysis
# ----------------------------------------------------------------------


class LiveExcludedTimeframeResponse(BaseModel):
    model_config = _PUBLIC

    timeframe: str
    reasons: list[str]


class LiveFingerprintResponse(BaseModel):
    model_config = _PUBLIC

    timeframe: str
    version: int


class LiveAnalysisResponse(BaseModel):
    """The existing Phase 8 analysis over confirmed, available candles."""

    model_config = _PUBLIC

    session_id: str
    provenance: Provenance
    market_currency: Currency
    market_as_of: str
    """Market time: the end of the latest confirmed candle analysed."""

    requested_at: str
    included: list[str]
    excluded: list[LiveExcludedTimeframeResponse]
    fingerprint: list[LiveFingerprintResponse]
    reused: bool
    """Served from the cache because its inputs are identical."""

    current: bool
    analysis: AnalysisResponse
    session: LiveSessionResponse


# ----------------------------------------------------------------------
# SSE
# ----------------------------------------------------------------------


class LiveEventEnvelope(BaseModel):
    """One Server-Sent Event's ``data``. Also the documented wire contract.

    ``HEARTBEAT`` means the transport is alive and carries the current cursor
    so a client can tell it missed something. It is not a candle and it
    touches no market state.
    """

    model_config = _PUBLIC

    protocol: Literal["viop.live.v1"] = "viop.live.v1"
    session_id: str
    event_id: int | None
    kind: Literal["TIMELINE", "STATE", "RESYNC_REQUIRED", "END", "HEARTBEAT"]
    server_time: str
    cursor: int
    entry: LiveTimelineEntryResponse | None = None
    session: LiveSessionResponse | None = None
    reason: str | None = None
